### PR TITLE
LL-3928 TypeScript migration

### PR DESCRIPTION
**Pull request by gre**
# Impact on end users of ledgerjs libraries

We designed the rework in a way that impact would be very low.

If you are using TypeScript, you should benefit the types of the library. If you were relying on FlowType: we have made a smooth transition curve where we would still support Flowtype defs for a few weeks (until we ourself fully migrate TypeScript in our projects) but we will eventually drop Flowtype completely (and make a major bump of ledgerjs)

The only type changes we did is on the Transport instance, we made it no longer have a parametric type,

so if you were writing code like

```
let transport: Transport<string> = Transport.open()
```

you can replace with

```
let transport: Transport = Transport.open()
```

---

We have also dropped the deprecated usage of 

```js
createPaymentTransactionNew(...multi params)
getWalletPublicKey(...multi params)
signP2SHTransaction(...multi params)
``` 

which should be replaced by the object style. It was console.warn as deprecated for 2 years so we expect it is not impacting much.

# TypeScript internal rework

## we used `flow-to-ts`
[GitHub - Khan/flow-to-ts: Convert flow code to typescript](https://github.com/Khan/flow-to-ts)
We have been using the codemod made by KhanAcademy to, more or less painlessly, convert our flow typed javascript files to typescript.
Most of the heavy lifting was done with the codemod, but we had to come back and fixes many types / errors.
For a quick usage of the codemod, here is how we used it
```bash
npx @khanacademy/flow-to-ts --prettier --write --delete-source --inline-utility-types src/**/*.js
```
We ran that script inside each `packages/*` folder (we decided to go modular and not run the script on the whole repo at once)

## Internal main changes
**./packages/cryptoassets/src/currencies.js**

```js

export type CryptoCurrencyObjMap<F> = $ObjMap<typeof cryptocurrenciesById, F>

export type CryptoCurrencyConfig<C> = CryptoCurrencyObjMap<(arg0: any) => C>

```

**./packages/hw-transport/src/Transport.js**

[TypeScript: Documentation - Classes](https://www.typescriptlang.org/docs/handbook/2/classes.html#type-parameters-in-static-members)
Cannot use Parametric Type in static members in Typescript.
Removed Descriptor in static members
Removed parametric type from Transport to make it simpler

**./packages/hw-app-algorand/src/Algorand.ts**

Added a `any` type to response in sign, as it can either be an object or an array from what I understand
```js
let response: any = null;
```

**./packages/hw-app-btc/src/Btc.ts**

Reworked without `arguments`  😱 and removing the deprecated calls
```js
getWalletPublicKey(
    path: string,
    opts?:
      | boolean
      | {
          verify?: boolean;
          format?: AddressFormat;
        }
  ): Promise<{
    publicKey: string;
    bitcoinAddress: string;
    chainCode: string;
  }> {
    let options;

    if (arguments.length > 2 || typeof opts === "boolean") {
      console.warn(
        "btc.getWalletPublicKey deprecated signature used. Please switch to getWalletPublicKey(path, { format, verify })"
      );
      options = {
        verify: !!opts,
        format: arguments[2] ? "p2sh" : "legacy",
      };
    } else {
      options = opts || {};
    }

    return getWalletPublicKey(this.transport, { ...options, path });
  }



createPaymentTransactionNew(arg: CreateTransactionArg) {
    if (arguments.length > 1) {
      console.warn(
        "@ledgerhq/hw-app-btc: createPaymentTransactionNew multi argument signature is deprecated. please switch to named parameters."
      );
      arg = fromDeprecateArguments(arguments, [
        "inputs",
        "associatedKeysets",
        "changePath",
        "outputScriptHex",
        "lockTime",
        "sigHashType",
        "segwit",
        "initialTimestamp",
        "additionals",
        "expiryHeight",
        "useTrustedInputForSegwit",
      ]);
    }

    return createTransaction(this.transport, arg);
  }


signP2SHTransaction(arg: SignP2SHTransactionArg) {
    if (arguments.length > 1) {
      console.warn(
        "@ledgerhq/hw-app-btc: signP2SHTransaction multi argument signature is deprecated. please switch to named parameters."
      );
      const [
        inputs,
        associatedKeysets,
        outputScriptHex,
        lockTime,
        sigHashType,
        segwit,
        transactionVersion,
      ] = arguments;
      arg = {
        inputs,
        associatedKeysets,
        outputScriptHex,
        lockTime,
        sigHashType,
        segwit,
        transactionVersion,
      };
      arg = fromDeprecateArguments(arguments, [
        "inputs",
        "associatedKeysets",
        "outputScriptHex",
        "lockTime",
        "sigHashType",
        "segwit",
        "transactionVersion",
      ]);
    }

    return signP2SHTransaction(this.transport, arg);
  }

```

**packages/hw-transport-node-ble/src/TransportNodeBle.ts**

I had to change the returnType because Promise<Buffer> would not work
Went for Promise<any> meanwhile
```js
exchange = (apdu: Buffer): Promise<Buffer> =>
    this.exchangeAtomicImpl(async () => {
      try {
        const msgIn = apdu.toString("hex");
        log("apdu", `=> ${msgIn}`);
        const data = await merge(
          this.notifyObservable.pipe(receiveAPDU),
          sendAPDU(this.write, apdu, this.mtuSize)
        ).toPromise();
        const msgOut = data.toString("hex");
        log("apdu", `<= ${msgOut}`);
        return data;
      } catch (e) {
        log("ble-error", "exchange got " + String(e));

        if (this.notYetDisconnected) {
          // in such case we will always disconnect because something is bad.
          await disconnectDevice(this.device).catch(() => {}); // but we ignore if disconnect worked.
        }

        throw e;
      }
    });

```

**packages/react-native-hid/package.json**
**pacjages/react-native-ble/package.json**

Removed the peerDependencies and moved them to dependencies because we did not find a solution to lint / build with peerDeps yet
- react-native@^0.63.4
- react-native-ble-plx@1.0.1

Apparently `rollup` can help us out with this issue.  We NEED to come back to this and fix it with peerDependencies

## Documentation
Remove the doc script from these breaking packages
- hw-transport-node-ble
- hw-transport-web-ble
- hw-transport-http
- hw-transport-mocker
- react-native-hw-transport-ble

## Tests
Changed the way BTC tests were written because they were still using the old API. Fixed the test to reflect the deprecation of the old APIs

### CODE REVIEW

File: packages/hw-transport-node-hid-singleton/src/listenDevices.ts [hunk 1]
```diff
@@ -1,34 +1,34 @@
...
+
 const monitor = () => {
   if (!monitoring) {
     monitoring = true;
     usbDetect.startMonitoring();
   }
-  return () => {};
+
+  return () => {
+    usbDetect.stopMonitoring();
```
Comment (gre) [inline packages/hw-transport-node-hid-singleton/src/listenDevices.ts:30]: ⚠️ logic has changed!

File: packages/hw-transport-node-hid-singleton/src/listenDevices.ts [hunk 2]
```diff
@@ -74,11 +73,8 @@ export const listenDevices = (
 
   usbDetect.on(addEvent, add);
   usbDetect.on(removeEvent, remove);
-
   return () => {
     if (timeout) clearTimeout(timeout);
-    usbDetect.off(addEvent, add);
-    usbDetect.off(removeEvent, remove);
```
Comment (gre) [inline packages/hw-transport-node-hid-singleton/src/listenDevices.ts:81]: ⚠️ logic has changed!

File: packages/hw-app-btc/src/createTransaction.ts [hunk 3]
```diff
@@ -77,14 +74,11 @@ export async function createTransaction(
...
-    useTrustedInputForSegwit,
     onDeviceStreaming,
     onDeviceSignatureGranted,
     onDeviceSignatureRequested,
-  } = {
-    ...defaultsSignTransaction,
-    ...arg,
-  };
+  } = { ...defaultsSignTransaction, ...arg };
+  let { useTrustedInputForSegwit } = { ...defaultsSignTransaction, ...arg };
```
Comment (gre) [inline packages/hw-app-btc/src/createTransaction.ts:81]: that part is weird. we should put the first object in a variable to not spread it twice.

File: packages/hw-app-btc/src/finalizeInput.ts [hunk 4]
```diff
@@ -1,37 +1,35 @@
...
   let offset = 0;
-  let p1 = 0x80;
+  const p1 = 0x80;
   const isDecred = additionals.includes("decred");
+
   ///WARNING: Decred works only with one call (without chunking)
   //TODO: test without this for Decred
   if (isDecred) {
-    return transport.send(0xe0, 0x4a, p1, 0x00, outputScript);
+    await transport.send(0xe0, 0x4a, p1, 0x00, outputScript);
```
Comment (gre) [inline packages/hw-app-btc/src/finalizeInput.ts:23]: ⚠️ the logic changed!! we used to return (meaning STOP the logic). whereas here we will not return and continue on the while.

File: packages/hw-app-eth/src/Eth.ts [hunk 5]
```diff
@@ -720,19 +768,19 @@ eth.signPersonalMessage("44'/60'/0'/0/0", Buffer.from("test").toString("hex")).t
    */
   starkSignTransfer(
     path: string,
-    transferTokenAddress?: string,
+    transferTokenAddress: string,
```
Comment (gre) [inline packages/hw-app-eth/src/Eth.ts:771]: this should be `| undefined`

File: packages/hw-app-eth/src/Eth.ts [hunk 6]
```diff
@@ -793,10 +846,10 @@ eth.signPersonalMessage("44'/60'/0'/0/0", Buffer.from("test").toString("hex")).t
    */
   starkSignTransfer_v2(
     path: string,
-    transferTokenAddress?: string,
+    transferTokenAddress: string,
```
Comment (gre) [inline packages/hw-app-eth/src/Eth.ts:849]: same

File: packages/hw-app-eth/src/Eth.ts [hunk 7]
```diff
@@ -793,10 +846,10 @@ eth.signPersonalMessage("44'/60'/0'/0/0", Buffer.from("test").toString("hex")).t
...
     transferQuantizationType: StarkQuantizationType,
-    transferQuantization?: BigNumber,
-    transferMintableBlobOrTokenId?: BigNumber,
+    transferQuantization: BigNumber,
+    transferMintableBlobOrTokenId: BigNumber,
```
Comment (gre) [inline packages/hw-app-eth/src/Eth.ts:852]: same?

File: packages/cryptoassets/src/exchange.ts [hunk 8]
```diff
@@ -1,18 +1,25 @@
...
 }
 
 /**
  *
  */
 export const findExchangeCurrencyConfig = (
   id: string
-): ?{ config: string, signature: string } => configs[id];
+):
+  | {
```
Comment (JunichiSugiura) [inline packages/cryptoassets/src/exchange.ts:19]: 🤔

File: packages/hw-app-btc/README.md [hunk 9]
```diff
@@ -214,9 +211,11 @@ Returns **[Buffer](https://nodejs.org/api/buffer.html)**
...
...cs/Web/JavaScript/Reference/Global_Objects/Number)?, sigHashType: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?, segwit: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?, initialTimestamp: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?, additionals: [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>, expiryHeight: [Buffer](https://nodejs.org/api/buffer.html)?, useTrustedInputForSegwit: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?, onDeviceStreaming: function (arg0: {progress: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), total: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), index: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)}): void?, onDeviceSignatureRequested: function (): void?, onDeviceSignatureGranted: function (): void?}
```
Comment (JunichiSugiura) [inline packages/hw-app-btc/README.md:214]: ⚠️

File: packages/hw-app-btc/package.json [hunk 10]
```diff
@@ -24,25 +24,22 @@
...
-  "devDependencies": {
-    "flow-bin": "^0.149.0"
-  },
   "scripts": {
-    "flow": "flow",
     "clean": "bash ../../script/clean.sh",
     "build": "bash ../../script/build.sh",
     "watch": "bash ../../script/watch.sh",
-    "doc": "bash ../../script/doc.sh"
+    "doc": "bash ../../script/doc-ts.sh"
```
Comment (JunichiSugiura) [inline packages/hw-app-btc/package.json:42]: remove `doc.sh` file and rename `doc-ts.sh`

File: packages/hw-app-btc/src/Btc.ts [hunk 11]
```diff
@@ -65,24 +63,17 @@ export default class Btc {
...
-        "btc.getWalletPublicKey deprecated signature used. Please switch to getWalletPublicKey(path, { format, verify })"
-      );
-      options = {
-        verify: !!opts,
-        format: arguments[2] ? "p2sh" : "legacy",
-      };
-    } else {
-      options = opts || {};
-    }
+    const options = opts || {};
```
Comment (JunichiSugiura) [inline packages/hw-app-btc/src/Btc.ts:75]: ⚠️

File: packages/hw-app-btc/src/Btc.ts [hunk 12]
```diff
@@ -128,31 +126,13 @@ export default class Btc {
...
-        "lockTime",
-        "sigHashType",
-        "segwit",
-        "initialTimestamp",
-        "additionals",
-        "expiryHeight",
-        "useTrustedInputForSegwit",
-      ]);
-    }
+  createPaymentTransactionNew(arg: CreateTransactionArg): Promise<string> {
```
Comment (JunichiSugiura) [inline packages/hw-app-btc/src/Btc.ts:135]: ⚠️

File: packages/hw-app-btc/src/Btc.ts [hunk 13]
```diff
@@ -169,58 +149,26 @@ btc.createTransaction({
...
-        "inputs",
-        "associatedKeysets",
-        "outputScriptHex",
-        "lockTime",
-        "sigHashType",
-        "segwit",
-        "transactionVersion",
-      ]);
-    }
+  signP2SHTransaction(arg: SignP2SHTransactionArg): Promise<string[]> {
```
Comment (JunichiSugiura) [inline packages/hw-app-btc/src/Btc.ts:158]: ⚠️

**Review by JunichiSugiura — COMMENTED** [review]
Reviewed carefully till `packages/hw-app-btc/src/createTransaction.ts` but phew it's sketchy.

(I should have commented warnings on removed lines.)

**Review by JunichiSugiura — COMMENTED** [review]
👀 till openTransportReplayer.ts

File: packages/hw-app-eth/tests/Eth.test.ts [hunk 14]
```diff
@@ -307,19 +301,21 @@ test("signAllowance", async () => {
...
     => e004000084058000002c8000003c800000000000000000000000f86d018504e3b2920082520894dac17f958d2ee523a2206206994597c13d831ec7872bd72a24874000b844095ea7b30000000000000000000000000102030405060708090a0b0c0d0e0f1011121314ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
     <= 1b3fa6a78fb25f87f063fc8db5cb4efc1794e01c973994e26a6fa1603c3ac3db9d3dc98795b5f99ba1eeae84ef01ecbfad188f00446d56b6e9a0eb9ec6f4bae7fe9000
     `)
   );
-  const transport = await Transport.open();
   const eth = new Eth(transport);
   const tokenInfo = byContractAddress(
     "0xdac17f958d2ee523a2206206994597c13d831ec7"
   );
+  await eth.provideERC20TokenInformation(tokenInfo as TokenInfo);
```
[the remark this answers is not in the document]
Comment (gre) [reply packages/hw-app-eth/tests/Eth.test.ts:318]: please remove it. it was a previous mistake.

File: packages/hw-app-polkadot/src/Polkadot.ts [hunk 15]
```diff
@@ -162,14 +169,16 @@ export default class Polkadot {
...
+
       if (
         returnCode === SW_ERROR_DATA_INVALID ||
         returnCode === SW_ERROR_BAD_KEY_HANDLE
       ) {
         const errorMessage = response
           .slice(0, response.length - 2)
           .toString("ascii");
-        throw new TransportError(errorMessage);
+        // FIXME: add correct id to TransportError
```
[the remark this answers is not in the document]
Comment (gre) [reply packages/hw-app-polkadot/src/Polkadot.ts:180]: I think you can just do 

```
        throw new TransportError(errorMessage, "Sign");
        ```

File: packages/hw-app-btc/src/getTrustedInput.ts [hunk 16]
```diff
@@ -46,28 +44,33 @@ export async function getTrustedInput(
...
+
   if (!outputs || !locktime) {
     throw new Error("getTrustedInput: locktime & outputs is expected");
   }
+
   const isDecred = additionals.includes("decred");
   const isXST = additionals.includes("stealthcoin");
 
-  const processScriptBlocks = async (script, sequence) => {
+  const processScriptBlocks = async (script, sequence?: Buffer) => {
```
Comment (JunichiSugiura) [inline packages/hw-app-btc/src/getTrustedInput.ts:55]: sequence is optional now?

File: packages/cryptoassets/src/exchange.ts [hunk 17]
```diff
@@ -1,18 +1,25 @@
...
+      config: string;
+
+      signature: string;
+    }
+  | null
+  | undefined => configs[id];
```
Comment (gre) [inline packages/cryptoassets/src/exchange.ts:25]: i don't understand the last `| undefined => configs[id]`

**Comment by gre** [comment]
We'll take this work and have a `rc` version of ledgerjs

**Pull request merged**